natlib/svcd.c

```diff
@@ -30,6 +31,63 @@ static const LUA_REG_TYPE svcd_meta_map[] =
 // Maintainer: Michael Andersen <michael@steelcode.com>
 /////////////////////////////////////////////////////////////
 
+//Lua: storm.n.svcd_advert_received(pay, srcip, srcport)
+//Dispatch an advertisment
+
+int svcd_advert_received(lua_State *L) 
+{
+	char *srcip = luaL_checkstring(L, 2);
+        printf("Service advertisment %s\n", srcip);
+	
+	//check if pay is nil, don't proceed if true
+	int nil_pay=lua_isnil(L,1);
+        if(nil_pay==1)
+	{
+ 		return 0;
+	}
+
+	char *pay = luaL_checkstring(L, 1);
+	int srcport = luaL_checknumber(L, 3);
+	lua_pushlightfunction(L, libmsgpack_mp_unpack);      
+       	lua_pushstring(L,pay); 
+	lua_call(L,1,1);	
+ 	lua_pushvalue(L, -1);
+    	lua_pushnil(L);
+    	while (lua_next(L, -2))
```

Comment:
We were able to do this key, value pair loop as a for loop which we thought was pretty elegant ([here](https://github.com/UCB-IoET/ioet_contrib/pull/31/files#diff-c26fa867b57c54ac1f31fbf7d0c078a6R199)).  From the looks of it your implementation is identical to the for loop version and not wrong but might be worth checking out.  I'm actually not sure which is more readable as they both rely on understanding how lua_next works